internal/icinga2/client.go

```diff
@@ -0,0 +1,388 @@
+		} else {
+			client.Logger.Errorw("Catching up the API failed", zap.Error(err), zap.Duration("duration", time.Since(startTime)))
+
+			select {
+			case <-ctx.Done():
+			case catchupEventCh <- &catchupEventMsg{error: err}:
+			}
+		}
+
+		cancel()
+		close(catchupEventCh)
+	}()
+
+	return catchupEventCh, cancel
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the catch-up-phase, requested by catchupPhaseRequest, events from the Event Stream API will
+// be cached until the catch-up-phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the eventDispatcherEventStream and catchupPhaseRequest channels.
+func (client *Client) worker() {
+	var (
+		// catchupEventCh either emits events generated during the catch-up-phase from catch-up-workers or one final
+		// error if something went wrong. It will be closed when catching up is done, which indicates the select below
+		// to switch phases. When this variable is nil, this Client is in the normal operating phase.
+		catchupEventCh chan *catchupEventMsg
+		// catchupCancel cancels, if not nil, all running catch-up-workers, e.g., when restarting catching-up.
+		catchupCancel context.CancelFunc
+
+		// catchupBuffer holds Event Stream events to be replayed after the catch-up-phase has finished.
+		catchupBuffer = make([]*event.Event, 0)
+		// catchupCache maps event.Events.Name to API time to skip replaying outdated events.
+		catchupCache = make(map[string]time.Time)
+
+		// catchupFailCounter indicates how many prior catch-up-phase attempts have failed. It will be used to
+		// rate limit catch-up-phase restarts.
+		catchupFailCounter int
+	)
+
+	// catchupReset resets all catchup variables to their initial empty state.
+	catchupReset := func() {
+		catchupEventCh, catchupCancel = nil, nil
+		catchupBuffer = make([]*event.Event, 0)
+		catchupCache = make(map[string]time.Time)
+	}
+
+	// catchupCacheUpdate updates the catchupCache if this eventMsg seems to be the latest of its kind.
+	catchupCacheUpdate := func(ev *eventMsg) {
+		ts, ok := catchupCache[ev.event.Name]
+		if !ok || ev.apiTime.After(ts) {
+			catchupCache[ev.event.Name] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.catchupPhaseRequest:
+			if catchupEventCh != nil {
+				client.Logger.Warn("Switching to catch-up-phase was requested while already catching up, restarting phase")
+				catchupCancel()
+			}
+
+			client.Logger.Info("Worker enters catch-up-phase, start caching up on Event Stream events")
+			catchupReset()
+			catchupEventCh, catchupCancel = client.startCatchupWorkers(
+				min(3*time.Minute, time.Duration(math.Exp2(float64(catchupFailCounter))-1)*time.Second))
```

Comment:
IMO the use of `math.Exp2` makes this harder to read than it has to be with the multiple casts required. As you don't need `catchupFailCounter` as an actual counter, you could consider something like a `var catchupDelay time.Duration` instead, where instead of of the `catchupFailCounter++`, you do something like this:

```go
// one-liner, maybe a the purpose of the max() in the middle is a bit unclear
catchupDelay = min(3*time.Minute, max(time.Second, 2*catchupDelay))

// expanded into a hopefully more readable if
if catchupDelay == 0 {
	catchupDelay = time.Second
} else {
	catchupDelay = min(3*time.Minute, 2*catchupDelay)
}
```

Reply:
Good idea. This is definitely a more readable approach.